### PR TITLE
Dev configure snap proxy during cloudinit

### PR DESCRIPTION
**Pull request by achilleasa**
## Description of change

This PR updates the cloud-init code-paths so that we can setup the snap store proxy while bootstrapping instead of relying on the proxyupdater worker to do that for us.  Unfortunately,
the worker runs at a later stage which means that if we need to pull any snaps before that happens (e.g. to install lxd or, in the future, install a mongo snap) we won't be able to use the snap proxy.

In addition, we now include support for setting up snap to use the store proxy automatically just by providing the proxy endpoint via the `snap-store-proxy-url` config setting. Juju can now directly contact the proxy to retrieve the required assertions and scan them to identify the appropriate store ID. To use this feature, the client initiating the bootstrap must be able to access the proxy. If that is not possible, the alternative is to manually provide the assertions and store ID values.

### 0) Prerequisite: install a snap store proxy

**Important**: replace **<domain>** with your host's **non-loopback** IP

```console
$ sudo snap install snap-store-proxy
$ docker run --name pg -e POSTGRES_PASSWORD=root -d postgres -p 5432:5432
$ sudo snap-proxy config proxy.domain="<domain>"
$ sudo snap-proxy create-database "postgresql://postgres:root@localhost:5432"
$ snap-proxy check-connections
http: https://dashboard.snapcraft.io: OK
http: https://login.ubuntu.com: OK
http: https://api.snapcraft.io: OK
postgres: localhost: OK
All connections appear to be accessible
$ sudo snap-proxy generate-keys
$ sudo snap-proxy register
# Record the store ID from the register command output

# Grab assertion list and store ID
$ curl -s http://<proxy-ip>/v2/auth/store/assertions > store-assertions.txt
$ cat store-assertions.txt | grep store: | cut -d' ' -f2 > store-id.txt
```

### 1) Set up snap proxy in already bootstrapped controller by manually specifying assertions

```console
$ juju bootstrap lxd test --no-gui
$ juju ssh -m controller 0 'sudo snap get core proxy'
Key          Value
proxy.http
proxy.https
proxy.store
$ echo "snap-store-proxy: \"`cat store-id.txt`\"\nsnap-store-assertions: |\n`cat store-assertions.txt|sed -e 's/^/  /g'`" > manual-conf.yaml
$ juju model-config -m controller manual-conf.yaml
# Wait a bit for proxyupdater to pick up the changes...
$ juju ssh -m controller 0 'sudo snap get core proxy'
Key          Value
proxy.http
proxy.https
proxy.store  $YOUR_STORE_ID$
```

### 2) Set up snap proxy in already bootstrapped controller by specifying just the proxy URL

```console
$ juju bootstrap lxd test --no-gui
$ juju ssh -m controller 0 'sudo snap get core proxy'
Key          Value
proxy.http
proxy.https
proxy.store
$ juju model-config -m controller snap-store-proxy-url="http://<proxy-ip>"
# Wait a bit for proxyupdater to pick up the changes...
$ juju ssh -m controller 0 'sudo snap get core proxy'
Key          Value
proxy.http
proxy.https
proxy.store  $YOUR_STORE_ID$
```

### 3) Set up snap proxy while bootstrapping by manually specifying assertions (airgapped proxy use-case)

```console
$ echo "snap-store-proxy: \"`cat store-id.txt`\"\nsnap-store-assertions: |\n`cat store-assertions.txt|sed -e 's/^/  /g'`" > manual-conf.yaml
$  juju bootstrap lxd test --no-gui --config manual-conf.yaml
$ juju ssh -m controller 0 'sudo snap get core proxy'
Key          Value
proxy.http
proxy.https
proxy.store  $YOUR_STORE_ID$
```

### 4) Set up snap proxy while bootstrapping by specifying the proxy URL

```console
# Run bootstrap with --debug
$  juju bootstrap lxd test --no-gui --debug --snap-store-proxy-url="http://<proxy-ip>"
....
Waiting for address
18:10:01 INFO  juju.cloudconfig.cloudinit cloudinit_ubuntu.go:332 auto-detected snap store assertions from proxy
18:10:01 INFO  juju.cloudconfig.cloudinit cloudinit_ubuntu.go:333 auto-detected snap store ID as "$YOUR_STORE_ID"
...

$ juju ssh -m controller 0 'sudo snap get core proxy'
Key          Value
proxy.http
proxy.https
proxy.store  $YOUR_STORE_ID$
```

## Documentation changes

We need to update our docs to:

- include a section about the "snap-store-proxy-url" config setting
- emphasize that in order for the above setting to work while **bootstrapping**, the client **must** be able to access the proxy (otherwise we can not pull the assertions from it). The alternative (e.g. air-gapped proxy) is to manually specify the assertion file/store ID config settings
- users don't need to provide any custom bootstrap commands to execute the steps from https://docs.ubuntu.com/snap-store-proxy/en/devices

### CODE REVIEW

**Comment by achilleasa** [comment]
!!build!!

(unrelated test failure)

**Comment by jameinel** [comment]
`$ docker run --name pg -e POSTGRES_PASSWORD=root -d postgres -p 5432:5432`
You can also 'snap install postgres' with something similar, which might be more consistent since you're doing a lot of snap stuff.

File: cloudconfig/cloudinit/cloudinit_centos.go [hunk 1]
```diff
@@ -183,11 +183,12 @@ func (cfg *centOSCloudConfig) getCommandsForAddingPackages() ([]string, error) {
 		cmds = append(cmds, addPackageMirrorCmd(cfg, newMirror))
 	}
 
+	pkgCmder := cfg.paccmder[jujupackaging.YumPackageManager]
```
Comment (SimonRichardson) [inline cloudconfig/cloudinit/cloudinit_centos.go]: This guaranteed to always exist?

File: cloudconfig/cloudinit/cloudinit_ubuntu.go [hunk 2]
```diff
@@ -300,13 +304,46 @@ func (cfg *ubuntuCloudConfig) addRequiredPackages() {
...
 		cfg.AddBootCmd(fmt.Sprintf(
 			`printf '%%s\n' %s > %s`,
-			utils.ShQuote(cfg.paccmder.ProxyConfigContents(proxySettings)),
+			utils.ShQuote(pkgCmder.ProxyConfigContents(aptProxy)),
 			filename))
 	}
+
+	// Write out the snap http/https proxy settings
+	if snapProxy := proxyCfg.SnapProxy(); (snapProxy != proxy.Settings{}) {
+		pkgCmder := cfg.paccmder[jujupackaging.SnapPackageManager]
```
Comment (SimonRichardson) [inline cloudconfig/cloudinit/cloudinit_ubuntu.go]: Guaranteed to exist?

File: cloudconfig/cloudinit/cloudinit_ubuntu.go [hunk 3]
```diff
@@ -300,13 +304,46 @@ func (cfg *ubuntuCloudConfig) addRequiredPackages() {
...
 		filename := config.AptProxyConfigFile
 		cfg.AddBootCmd(fmt.Sprintf(
 			`printf '%%s\n' %s > %s`,
-			utils.ShQuote(cfg.paccmder.ProxyConfigContents(proxySettings)),
+			utils.ShQuote(pkgCmder.ProxyConfigContents(aptProxy)),
 			filename))
 	}
+
+	// Write out the snap http/https proxy settings
+	if snapProxy := proxyCfg.SnapProxy(); (snapProxy != proxy.Settings{}) {
```
Comment (SimonRichardson) [inline cloudconfig/cloudinit/cloudinit_ubuntu.go]: Not really a fan of this `(snapProxy != proxy.Settings{})` - I guess a small tiny comment about checking if the `snapProxy` is not empty.

File: cloudconfig/cloudinit/cloudinit_centos.go (hunk 1 above)
Comment (achilleasa) [reply cloudconfig/cloudinit/cloudinit_centos.go]: Yes, we set it up here: https://github.com/juju/juju/pull/10602/files/321e9305cdfba10805492f9e7e5aa5077f2c2663#diff-69cc37668e4f4d65f4b1dea9a62219abR439

File: cloudconfig/cloudinit/cloudinit_ubuntu.go (hunk 3 above)
Comment (achilleasa) [reply cloudconfig/cloudinit/cloudinit_ubuntu.go]: Yeah, it was already there so I decided to keep it. Fun fact: if you remove the parentheses it does not compile!

**Comment by achilleasa** [comment]
$$merge$$

**Pull request merged**